### PR TITLE
[libra framework] stronger invariants on sent/received events

### PR DESCRIPTION
**Pull request by sblackshear**
Implementing the Rosetta API for Libra has revealed a few corner cases in our events. This PR overhauls the event semantics as follows:

1. Every `SentPaymentEvent { amount, currency_code }` emitted to stream *H* debits *H*.`address`'s `currency_code` balance by `amount`
2. Every `ReceivedPaymentEvent { amount, currency_code }` emitted to stream *H* credits *H*.`address`'s `currency_code balance` by `amount`
3. In almost all cases, a `SentPaymentEvent { amount, currency code, payee }` emitted to stream *H1* is coupled with a `ReceivedPaymentEvent { amount, currency_code, payer }` emitted to a stream *H2* such that *H1*.`address` = `payer` and *H2*.`address` = `payee`.

The only exceptions to rule 3 are minting, preburning, cancelling a burn, burning, and transaction fees.
- Minting creates funds from nowhere and credits a balance under address *A*. We represent this with a `ReceivedPaymentEvent` emitted to *H(A)* where payer = `0x0...0` (the reserved address). There is no corresponding `SentPaymentEvent`.
- Preburning debits a balance under address *A* and credits a `Preburn` resource under address *A*. We represent this using a `SentPaymentEvent` emitted to *H(A)* where the payee is *A* + a `PreburnEvent` where the `preburn_address` is *A*. There is no corresponding `ReceivedPaymentEvent`.
- Cancelling a burn debits a `Preburn` resource under address *A* and credits the balance of *A*.  We represent this using a `ReceivedPaymentEvent` emitted to *H(A)* where the payer is *A* + a `CancelBurnEvent` where the `preburn_address` is *A*.
- Burning debits a `Preburn` resource under address *A* and destroys the funds. We represent this using a `SentPaymentEvent emitted` to *H(A)* where the payee is the reserved address `0x0...0` + a `BurnEvent` where the `preburn_address` is *A*
- Finally, transaction fees debit a balance under *A* and credit a `TransactionFee` resource under `0xA550C18`. This emits no events.

The basic principles here are that the reserved address `0x0...0` as a payer or payee in an event means "no one", and that the non-balance resources that hold coins (the `Preburn` resource and the `TransactionFee` resource) are represented by the address that stores the resource.

Based on all of this, the rules for reconciling balances and events in Rosetta would be:

- See `SentPaymentEvent { amount, currency_code, payee }` emitted to stream *H*: debit *H*.`address`'s `currency_code` balance by `amount`. Ignore `payee`.
- See `ReceivedPaymentEvent { amount, currency_code, payer }` emitted to stream *H*: credit *H*.`address`'s `currency_code` balance by `amount`. Ignore `payer`.
- See executed transaction sent from *A* with `gas_used` = X and `gas_price` = Y: debit *A*'s balance by X * Y and credit `0xA550C18`'s balance by X * Y.
- See `MintEvent { amount, currency_code }`: increase quantity of `currency_code` in the system by amount.
- See `BurnEvent { amount, currency_code }`: decrease quantity of `currency_code` in the system by amount.

The meat of this change is in `LibraAccount`; almost everything else is changing tests to adapt to the new and more restrictive APIs. I also deleted several mvir tests that no longer make sense now that we cannot withdraw a coin into a local.

### CODE REVIEW

File: language/stdlib/modules/LibraAccount.move [hunk 1]
```diff
@@ -307,11 +318,33 @@ module LibraAccount {
...
     // Withdraw `amount` Libra<Token> from the account balance under
     // `cap.account_address`
-    public fun withdraw_from<Token>(cap: &WithdrawCapability, amount: u64): Libra<Token>
-    acquires Balance, AccountOperationsCapability {
-        let account_balance = borrow_global_mut<Balance<Token>>(cap.account_address);
-        // The sender has retained her withdraw privileges--proceed.
-        withdraw_from_balance<Token>(cap.account_address, account_balance, amount)
+    fun withdraw_from<Token>(
+        cap: &WithdrawCapability,
+        payee: address,
```
Comment (tzakian) [inline language/stdlib/modules/LibraAccount.move:323]: This is going to make account limits and making sure we don't count intra-vasp transfers so much easier as well!

File: language/stdlib/modules/LibraAccount.move [hunk 2]
```diff
@@ -157,44 +157,54 @@ module LibraAccount {
...
+        let coin2_balance = balance<Coin2>(cap_address);
+        // use the LBR reserve address as `payee_address`
+        let payee_address = LBR::reserve_address();
+        let coin1 = withdraw_from<Coin1>(cap, payee_address, coin1_balance, x"");
+        let coin2 = withdraw_from<Coin2>(cap, payee_address, coin2_balance, x"");
+        // Create `amount_lbr` LBR
+        let (lbr, coin1, coin2) = LBR::create(amount_lbr, coin1, coin2);
+        // use the reserved address as the payer for the LBR payment because the funds did not come
+        // from an existing balance
+        deposit(CoreAddresses::VM_RESERVED_ADDRESS(), cap_address, lbr, x"", x"");
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/LibraAccount.move:180]: The discipline I'm going for here is:
- If the funds were created from scratch (i.e., minted), the sender is 0x0
- If the funds were debited from an existing resource, the sender is the address this resource is stored under

In this case, the funds are created from scratch, so 0x0 is the right sender.

File: language/stdlib/modules/LibraAccount.move [hunk 3]
```diff
@@ -157,44 +157,54 @@ module LibraAccount {
...
+        // TODO: eliminate these self-deposits by withdrawing appropriate amounts up-front
+        // Deposit the Coin1/Coin2 remainders
+        deposit(cap_address, cap_address, coin1, x"", x"");
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/LibraAccount.move:183]: Yeah, writing this in a way that puts as much logic as possible in `LBR` is an interesting challenge... I like the mutable refs idea.

**Pull request closed without merging**